uplink/storage/segments/store.go

```diff
@@ -218,179 +184,77 @@ func (s *segmentStore) Get(ctx context.Context, path storj.Path) (rr ranger.Rang
-				ErasureShareSize: int32(rs.ErasureShareSize()),
-			},
-			RootPieceId:  pieceID,
-			RemotePieces: remotePieces,
-		},
-		SegmentSize:    readerSize,
-		ExpirationDate: expiration,
-		Metadata:       metadata,
+		return rr, info.SegmentEncryption, nil
 	}
-	return pointer, nil
 }
 
 // Delete requests the satellite to delete a segment and tells storage nodes
 // to delete the segment's pieces.
-func (s *segmentStore) Delete(ctx context.Context, path storj.Path) (err error) {
+func (s *segmentStore) Delete(ctx context.Context, streamID storj.StreamID, segmentIndex int32) (err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	bucket, objectPath, segmentIndex, err := splitPathFragments(path)
-	if err != nil {
-		return err
-	}
-
-	limits, privateKey, err := s.metainfo.DeleteSegment(ctx, bucket, objectPath, segmentIndex)
+	segmentID, limits, privateKey, err := s.metainfo.BeginDeleteSegment(ctx, metainfo.BeginDeleteSegmentParams{
+		StreamID: streamID,
+		Position: storj.SegmentPosition{
+			Index: segmentIndex,
+		},
+	})
 	if err != nil {
 		return Error.Wrap(err)
 	}
 
-	if len(limits) == 0 {
-		// inline segment - nothing else to do
-		return
+	if len(limits) != 0 {
+		// remote segment - delete the pieces from storage nodes
+		err = s.ec.Delete(ctx, limits, privateKey)
+		if err != nil {
+			return Error.Wrap(err)
+		}
 	}
 
-	// remote segment - delete the pieces from storage nodes
-	err = s.ec.Delete(ctx, limits, privateKey)
+	err = s.metainfo.FinishDeleteSegment(ctx, metainfo.FinishDeleteSegmentParams{
+		SegmentID: segmentID,
+		// TODO add delete results
+	})
 	if err != nil {
 		return Error.Wrap(err)
 	}
 
 	return nil
 }
 
-// List retrieves paths to segments and their metadata stored in the metainfo
-func (s *segmentStore) List(ctx context.Context, prefix, startAfter, endBefore storj.Path, recursive bool, limit int, metaFlags uint32) (items []ListItem, more bool, err error) {
-	defer mon.Task()(&ctx)(&err)
-
-	bucket, strippedPrefix, _, err := splitPathFragments(prefix)
-	if err != nil {
-		return nil, false, Error.Wrap(err)
-	}
-
-	list, more, err := s.metainfo.ListSegments(ctx, bucket, strippedPrefix, startAfter, endBefore, recursive, int32(limit), metaFlags)
-	if err != nil {
-		return nil, false, Error.Wrap(err)
-	}
-
-	items = make([]ListItem, len(list))
-	for i, itm := range list {
-		items[i] = ListItem{
-			Path:     itm.Path,
-			Meta:     convertMeta(itm.Pointer),
-			IsPrefix: itm.IsPrefix,
-		}
-	}
-
-	return items, more, nil
-}
-
 // CalcNeededNodes calculate how many minimum nodes are needed for download,
 // based on t = k + (n-o)k/o
-func CalcNeededNodes(rs *pb.RedundancyScheme) int32 {
+func CalcNeededNodes(rs storj.RedundancyScheme) int32 {
 	extra := int32(1)
 
-	if rs.GetSuccessThreshold() > 0 {
-		extra = ((rs.GetTotal() - rs.GetSuccessThreshold()) * rs.GetMinReq()) / rs.GetSuccessThreshold()
+	if rs.OptimalShares > 0 {
+		extra = int32(((rs.TotalShares - rs.OptimalShares) * rs.RequiredShares) / rs.OptimalShares)
```

Comment:
https://storjlabs.atlassian.net/browse/V3-2485
For download, I am convinced that TotalShares should be removed from the formula. TotalShares are theoretical but never existing shares, just needed to query enough nodes for upload.

Reply:
I would prefer to make such change as a separate PR

Reply:
However, I saw that you touched that and only want point to V3-2485.